### ISSUE TITLE
CDFJ runs on hardware, not Stella...

### ISSUE THREAD

**Issue by andrew-davie** [stella-emu/stella#883]
I have a CDFJ binary which starts OK on NTSC hardware via Harmony cart, but shows corrupt graphics screen and crash on Stella 6.6. I don't want to share the binary publicly, but anyone on the Stella team wanting to debug/diagnose this issue please PM me for a copy.  The binary runs OK in Gopher.

**Comment by andrew-davie** [comment]
I've made a stand-alone binary that shows the same symptoms but doesn't include all the other stuff i didn't want to share. The pictures show the two screens that should show. Again, working fine on hardware but not Stella 6.6, which shows the corrupt frame/screen shown.

NTSC hardware screenshots showing expected visuals on 2 screens...

![IMG20220329164545](https://user-images.githubusercontent.com/16424075/160550752-56595903-347c-4467-9b63-a91df1367cf8.jpg)

![IMG20220329164548](https://user-images.githubusercontent.com/16424075/160550798-12f11f32-3e99-4960-ae50-4f442a822853.jpg)

Binary to test...
[CDFJ_fail.bin.zip](https://github.com/stella-emu/stella/files/8368779/CDFJ_fail.bin.zip)


Stella screenshot.  Mostly this, but occasionally traps to debugger...
<img width="752" alt="Screen Shot 2022-03-29 at 4 51 49 pm" src="https://user-images.githubusercontent.com/16424075/160551040-4cd9c910-ad45-4560-b0c8-c72c0cac6761.png">

**Comment by andrew-davie** [comment]
Notes:

* have tried Stella with and without Developer Settings - no difference - never works
* have tried Gopher with and without startup memory/pin randomisation - always works
* always works on hardware in all tests so far

**Comment by thrust26** [comment]
How much RAM are you using?

**Comment by DirtyHairy** [comment]
FWIW it works in Stellerator, which uses thumbulator, too.

**Comment by andrew-davie** [comment]
> How much RAM are you using?


Not much, really. In the latest binary which has even more stuff removed (and which fails in the same way)...

```
   text    data     bss     dec     hex filename
      0       0       0       0       0 sound.o
    104       0       0     104      68 custom.o
      0       0       0       0       0 main.o
      0       0       0       0       0 attribute.o
      0       0       0       0       0 decodecaves.o
      0       0       0       0       0 characterset.o
      0       0       0       0       0 cavedata.o
      0       0       0       0       0 drawplayer.o
      0       0       0       0       0 bitshapes.o
      0       0       0       0       0 overlay.o
      0       0       0       0       0 drawscreen.o
      0       0       0       0       0 circleSwipe.o
      0       0       0       0       0 scoreLine.o
      0       0       0       0       0 animations.o
      0       0       0       0       0 menu.o
      0       0       0       0       0 atarivox.o
   6812      44     448    7304    1c88 main/bin/armcode.elf
```




Essentially about 14 "strips" at 198 bytes each are the main RAM usage here.

**Comment by andrew-davie** [comment]
> Works fine in the latest GIT build of Stella when 'RAM randomization' is disabled.
> 
> I can test on 6.6 proper later today, when I get to my dev machine.

OK, confirmed here too. Sorry, when I chose player/developer I did not realise that RAM randomisation remained checked.
Issue can be closed. I will report back when I find the exact culprit here, but obviously an uninitialised var in my code.
Thanks for looking at it.

**Comment by DirtyHairy** [comment]
>  but obviously an uninitialised var in my code.

Or a bad pointer or an array overflow.

**Comment by andrew-davie** [comment]
OK, I promised to return. I now have my code functioning on hardware. The exact sequence of events was tricky to diagnose because, like, spaghetti code.  Essentially the scanline jump table for drawing a screen was uninitialised because a jump vector table was using an uninitialised (random) value which worked when memory was zero, but mostly not when non-zero.  The Gopher emulator has been updated to include randomising all RAM - when I was testing this earlier, I did not realise that the cruicial part of RAM triggering this bug/crash was in fact always zero in that emulator.  On the plus side, bugs are great for improving emulator efficiency. I just had a doozy where both Stella and Gopher performed perfectly, but hardware had significant glitches -- caused, I'm uncomfortably unsure, by a RAM stomping.  But anyway, the irony on THIS issue for me is that even though my hardware ran the binary perfectly, as did Gopher, but Stella crashed... is that Stella was in fact correct... I just got lucky on my hardware's behaviour, causing it not to crash, and on Gopher's non-randomisation of a crucial memory location, causing **it** not to crash.
